pkg/kubelet/gpu/nvidia/nvidia.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nvidia
+
+import (
+	"sync"
+
+	gpuutil "k8s.io/kubernetes/pkg/kubelet/gpu/nvidia/util"
+	gputypes "k8s.io/kubernetes/pkg/kubelet/gpu/types"
+)
+
+var (
+	gpuMutex sync.Mutex
+)
+
+// TODO: If use NVML in the future, the implementation could be more complex,
+// and more powerful!
+
+type NvidiaGPU struct {
+	gpuInfo []gputypes.GPUDevice
+}
+
+func ProbePlugin() (gputypes.GPUPlugin, error) {
+	var nvidiaGPU NvidiaGPU
+
+	err := nvidiaGPU.InitPlugin()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &nvidiaGPU, nil
+}
+
+// Get the vendor information.
+func (nvidiaGPU *NvidiaGPU) Vendor() string {
+	return gpuutil.Vendor
+}
+
+func (nvidiaGPU *NvidiaGPU) InitPlugin() error {
+	gpuMutex.Lock()
+	defer gpuMutex.Unlock()
+
+	if err := gpuutil.Init(); err != nil {
+		return err
+	}
+
+	allPaths, err := gpuutil.Discovery()
+
+	if err != nil {
+		return err
+	}
+
+	// Initialize the Nvidia device information.
+	for _, path := range allPaths {
+		nvidiaGPU.gpuInfo = append(nvidiaGPU.gpuInfo, gputypes.GPUDevice{Path: path, Status: gputypes.GPUFree, ContainerID: ""})
+	}
+
+	return nil
+}
+
+// Release the Plugin, could be useful once we use NVML.
+func (nvidiaGPU *NvidiaGPU) ReleasePlugin() error {
+	gpuMutex.Lock()
+	defer gpuMutex.Unlock()
+
+	nvidiaGPU.gpuInfo = nil
+
+	return nil
+}
+
+func (nvidiaGPU *NvidiaGPU) Capacity() (int, error) {
+	gpuMutex.Lock()
+	defer gpuMutex.Unlock()
+
+	return len(nvidiaGPU.gpuInfo), nil
+}
+
+func (nvidiaGPU *NvidiaGPU) AvailableGPUs() int {
+	gpuMutex.Lock()
+	defer gpuMutex.Unlock()
+
+	freeGPUs := 0
+	for _, gpuItem := range nvidiaGPU.gpuInfo {
+		if gpuItem.Status == gputypes.GPUFree {
+			freeGPUs += 1
+		}
+	}
+
+	return freeGPUs
+}
+
+func (nvidiaGPU *NvidiaGPU) AllocateGPU(number int) (allocPaths []string) {
+	avaliableGPUs := nvidiaGPU.AvailableGPUs()
+
+	gpuMutex.Lock()
+	defer gpuMutex.Unlock()
+
+	if avaliableGPUs < number {
+		return
+	}
+
+	for _, gpuItem := range nvidiaGPU.gpuInfo {
```

Comment:
@Hui-Zhi Yes, I know,  I mean the "number" didn't play a role in limit the number of the "allocPaths". Assume that the "availableGPUs" is 4, the "nvidiaGPU.gpuInfo" array contains 4 GPUs, and all of the status of GPU are "Free", the "number" arguments is 2, judging from the current code, the "allocPaths" value the func returned will be 4, not 2, isn't it? If I wrong, pls correct me, thanks.